client/src/shared/state/Inventory.js

```diff
@@ -104,16 +107,155 @@ class Inventory {
         if (!ItemTypes[itemConfig.typeCode].hasUseEffect) return;
 
         this.hotbar.push(itemConfig);
-
+        this.keyToSlotIndex[this.hotbar.length - 1] = itemConfig.slotIndex;
         PubSub.publish(HOTBAR_ITEM);
         PubSub.publish(MODIFY_INVENTORY_ITEM);
+        if (this.saveSession === true) {
+            this.saveHotbar();
+        }
+    }
+
+    resetMapping() {
+        Utils.warning("This is a bug found or a localStorage mismatch");
+        this.keyToSlotIndex = {};
+        this.saveHotbar();
+        this.hotbar = [];
+        this.defaultHotBar();
     }
 
-    removeFromHotbar(itemConfig) {
+    updateSlotIndex(slotIndex) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (this.keyToSlotIndex[key] >= slotIndex) {
+                const updateValue = this.keyToSlotIndex[key] - 1;
+                if (updateValue >= 0) {
+                    this.keyToSlotIndex[key] = updateValue;
+                    if (this.saveSession === true) {
+                        this.saveHotbar();
+                    }
+                }
+            }
+        });
+    }
+
+    updateMapHotBarKeys(evictKey) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (key >= evictKey) {
+                const updateKey = key - 1;
+                if (updateKey >= 0) {
+                    this.keyToSlotIndex[updateKey] = this.keyToSlotIndex[key];
+                    delete this.keyToSlotIndex[key];
+                }
+            }
+        });
+    }
+
+    updateMapSlotIndex(evictValue) {
+        Object.keys(this.keyToSlotIndex).forEach((key, index) => {
+            if (this.keyToSlotIndex[key] >= evictValue) {
+                const newValue = this.keyToSlotIndex[key] - 1;
+                if (newValue >= 0) {
+                    this.keyToSlotIndex[key] = newValue;
+                }
+            }
+        });
+    }
+
+    removeFromHotbar(itemConfig, removedInventory = false) {
+        const getKeys = Object.keys(this.keyToSlotIndex);
+        const evictKey = getKeys.find((key) => this.keyToSlotIndex[key] === itemConfig.slotIndex);
+        const evictValue = this.keyToSlotIndex[evictKey];
+        if (typeof evictKey === "undefined") {
+            this.updateSlotIndex(itemConfig.slotIndex);
+            return;
+        }
+        delete this.keyToSlotIndex[evictKey];
+        this.updateMapHotBarKeys(evictKey);
+        if (removedInventory) {
+            this.updateMapSlotIndex(evictValue);
+        }
         this.hotbar = this.hotbar.filter((eachItem) => eachItem !== itemConfig);
 
         PubSub.publish(HOTBAR_ITEM);
         PubSub.publish(MODIFY_INVENTORY_ITEM);
+        if (this.saveSession === true) {
+            this.saveHotbar();
+        }
+    }
+
+    saveHotbar() {
+        const sessionIDMap = "saveHotBar";
```

Comment:
You have duplicate code (see line 215) - this means this constant should be saved as an `Inventory` class field so you can access it from all class functions. This would make it easier to change the sessionIDMap string later if they wanted to. Also, just nitpicking, but I would probably use a more descriptive variable name rather than `sessionIDMap` so people reading the code know what its purpose is. Maybe something like `HotBarStorageKey`